accepted/00065-salt-minions-update-cycle.md

```diff
@@ -0,0 +1,127 @@
+
+![Layout](images/00065-layout.png)
+
+_Figure 1_
+
+Process update or installation flow as follows:
+1. One RPM general package of Salt Minion **(1)** is provided for all possible platforms and architectures and is installed on PyPi server. However, Salt Minion is not _installed_ per se in the sense as for end-user, but as a PyPi package into the PyPi repository location. This way patches are still tracked at RPM `.spec` level.
+2. General statically compiled Python vanilla environment is provided as an RPM package from Uyuni repository and is allowed to be installed **(2)** on a Managed System, e.g. in `/opt/python3.8` or whatever version is needed. That way it is also allowed to have many Python versions as well as its modifications, if ever needed.
+3. Once virtual environment (not shown on the _Figure 1_) is created for the the separate Python interpreter, PIP is bootstrapping the entire Salt Minion against Uyuni-served PyPi server with all the required Salt Minion dependencies.
+
+This should result into running fully featured Salt Minion of latest version in an isolated virtual environment on an OS-isolated Python interpreter, which is aligned with the current Uyuni Server. Once Uyuni Server is updated to another version, so is the PyPi container image is also upgraded. Thus, Uyuni Server should issue mass-upgrade to all Salt Minions that should PIP-upgrade themselves, bumping up to the latest available version.
+
+Consequently, this mechanism will allow to keep always aligned Salt Minions across the entire infrastructure consistent and identical.
+
+## Fail-proof Updates
+
+Having running Salt Minion inside a virtual environment, allows to deliver fail-proof updates. The basic mechanism is to clone Python Virtual Environment from the current active Salt Minion and update the Salt Minion there.
+
+If update is finished successfully, Salt Minion is restarted into a new environment, which is set as default. Previos environment is kept, and is purged over one update cycle (one after current update). This way administrator always has an ability to roll-back to the previous known working version.
+
+If update failed, Salt Minion keeps running without changes, report is created and new Virtual Environment is purged immeditately.
+
+As shown on the _Figure 2_, the difference between RPM-based updates is that the software components are overwritten, instead of installed nearby. Therefore once Salt Minion is broken on upgrade for whatever reasons, it might be no longer possible to revert this with the faulty Salt Minion itself.
+
+![Layout](images/00065-ve.png)
+
+_Figure 2_
+
+According to the _Figure 2_, if updated Salt Minion is successfully updated and is trusted enough, then the statuses of each virtual environment are shifted, and "Previous Salt Minion" is purged.
+
+Naturally, the same way it is possible to update the entire Python version on production and roll-back to the previous Python version, in case things go wrong.
```

Comment:
Yes. Something nice of this idea is that rm -fr /opt will not break the system. You can bootstrap it again.

This makes me wonder, in your idea, why to detach the python binary from the salt-minion code? Deploy all together as salt-bin is doing, and you avoid the risk of failing during the creation of the venv on the site.

Reply:
I am not sure, are you actually fully understand how PyInstaller works? It does not installs any Python. It is a C-written bootstrap program-carrier, which sets up a temporary virtual environment, reads itself and, depending on the form of the distribution (directory or one file) loads the modules into Python interpreter shipped as a library.

You also keep missing the fact that PyInstaller does not bundle `libc`, which is not any kind of static distribution. To make it actually static, you would need to use something like [StaticX](https://github.com/JonathonReinhart/staticx) including libc.

So we basically have a choice:

a) Generate RPMs for each arch with all the static libraries inside, deps and everything and support all that.

Or:

b) Let every machine pip-install to build a Minion in gentoo-like fashion, completely cutting all the packaging overhead from a).

Reply:
I indeed do not dig into PyInstaller, and I did not asked about it. I am not convinced that is not silver-bulled, if this counts.

What I asked is that, if you are willing to do the job and provide an static Python, why do you need now a venv? You can install this static Python in /opt/salt-x.y and all the dependencies (python and non-python .so, .a, whatever) in this namespace (/opt-salt-x.y/lib, /opt-salt-x.y/lib/site-packages, /opt-salt-x.y/usr, /opt-salt-x.y/bin, etc, etc)

> You also keep missing the fact that PyInstaller does not bundle libc

I am not sure, as I did not discussed here anything about PyInstaller. But I think that this is not a problem, no?

Go generate static binaries, and rust binaries link to `libc`, `libthread` and so on, and what?

How many bugs can you expect from this? Is the proposal generate a single binary per arch, and forget about the version of the OS? So for all and each binary `ldd` will be empty?

> a) Generate RPMs for each arch with all the static libraries inside, deps and everything and support all that.

> b) Let every machine pip-install to build a Minion in gentoo-like fashion, completely cutting all the packaging overhead from a).

I am clearly misinterpreting something, as this b) option will require from the client to have GCC, python-devel, openssh-devel and many other packages before the venv can be build a la gentoo.

So I guess that you prefer the option a) here, but in that case no venv/pip is required.

Reply:
> What I asked is that, if you are willing to do the job and provide an static Python, why do you need now a venv? You can install this static Python in /opt/salt-x.y and all the dependencies (python and non-python .so, .a, whatever) in this namespace (/opt-salt-x.y/lib, /opt-salt-x.y/lib/site-packages, /opt-salt-x.y/usr, /opt-salt-x.y/bin, etc, etc)

https://github.com/uyuni-project/uyuni-rfc/blob/isbm-salt-minion-update-cycle/accepted/00065-salt-minions-update-cycle.md#fail-proof-updates

Reply:
> a) Generate RPMs for each arch with all the static libraries inside, deps and everything and support all that.

> b) Let every machine pip-install to build a Minion in gentoo-like fashion, completely cutting all the packaging overhead from a).

You're conveniently leaving out the path that `salt-bin` is actually taking if I understand things correctly: Using the oldest libc that you need to support when you build the PyInstaller "package". This isn't going to 100% guarantee that things will work, but chances are high that a single binary would be pretty portable from RHEL 6 up to SLES 15 and RHEL 8 (as long as you are on the same HW architecture).

Reply:
The `salt-bin` is just a [wrapper script](https://github.com/saltstack/salt-bin/blob/master/build.py#L126) around [PyInstaller](http://www.pyinstaller.org) providing [pre-defined spec](https://github.com/saltstack/salt-bin/blob/master/build.py#L14-L47), so let's just focus on what PyInstaller does. And the [bootstrap](https://github.com/pyinstaller/pyinstaller/tree/develop/bootloader) loader from PyInstaller does not provide `libc`, consequently making the resulting binary dynamic.

That said, to ensure Python environment works 100% on RHEL6, SLES11 etc, the `libc` should be also brought together. Granted, using [StaticX](https://github.com/JonathonReinhart/staticx) together with PyInstaller would include `libc`, but that is even worse hack.